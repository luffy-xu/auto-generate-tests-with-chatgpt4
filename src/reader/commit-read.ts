import { execSync } from 'child_process';
import fs from 'fs';

const filterEmpty = (line: string) => line.trim().length > 0;
enum EFileDiffType {
  ADD = 'add',
  MODIFY = 'modify',
  DELETE = 'delete',
  RENAME = 'rename',
}
export class CommitRead {
  private read(diffFile?: string) {
    if (diffFile && fs.existsSync(diffFile)) {
      return fs.readFileSync(diffFile).toString();
    }
    return this.readFromGit();
  }

  private readFromGit() {
    const files = execSync(
      'git diff --staged --ignore-all-space --diff-algorithm=minimal --function-context --no-ext-diff --no-color',
    ).toString();
    return files;
  }

  private parseFile(file: string) {
    const lines = file.split('\n');
    const [source, target] = lines[0].split(' ').filter(filterEmpty);
    const addLines = lines.filter((line) => line.startsWith('+')).length;
    const delLines = lines.filter((line) => line.startsWith('-')).length;

    let trimLines = 4;
    let mode = EFileDiffType.MODIFY;
    const line1 = lines[1];
    if (line1.startsWith('new file mode')) {
      mode = EFileDiffType.ADD;
      trimLines = 5;
    }

    if (line1.startsWith('deleted file mode')) {
      mode = EFileDiffType.DELETE;
      trimLines = 5;
    }

    if (line1.startsWith('similarity index')) {
      mode = EFileDiffType.RENAME;
      trimLines = 6;
    }

    return {
      source: source.slice(1),
      filename: target.slice(1),
      mode,
      diff: lines.slice(trimLines).join('\n').replace( /@@\s-.+\s\+.+\s@@\s+/, ''),
      add: addLines,
      del: delLines,
    }
  }

  getFileList(diffFile?: string) {
    const fileString = this.read(diffFile);
    const files = ('\n' + fileString).split('\ndiff --git ').filter(filterEmpty);
    return files.map((file) => this.parseFile(file));
  }
}
