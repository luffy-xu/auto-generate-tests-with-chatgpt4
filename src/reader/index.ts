import fs from 'fs';
import ora from 'ora';
import path from 'path';
import { userOptions } from 'src/constant';
import { IReadFileResult, ReadTypeEnum } from 'src/types';

import ReadTestFilePathsByDirectory from './reader-directory';
import StagedFileReader from './reader-git-stage';

class ReadFiles {
  private dirPath: string;
  private fileExtensions: string[];

  constructor({
    dirPath = userOptions.readFilesRoot,
    fileExtensions = userOptions.readFilesExtensions,
  } = {}) {
    this.dirPath = dirPath;
    this.fileExtensions = fileExtensions;
  }

  readTypeMap: Record<ReadTypeEnum, () => IReadFileResult[]> = {
    [ReadTypeEnum.Directory]: () => this.getTestFilePathByDir(),
    [ReadTypeEnum.GitStage]: () => this.getTestFilePathByGit(),
  };

  // Get all file paths by directory
  private getTestFilePathByDir(): IReadFileResult[] {
    const reader = new ReadTestFilePathsByDirectory();
    return reader.getDirFiles(this.dirPath);
  }

  // Get all file paths by git stage
  private getTestFilePathByGit(): IReadFileResult[] {
    const reader = new StagedFileReader();
    return reader.getStagedFiles();
  }

  // Check if a file has a valid extension
  private hasValidExtension(file: string): boolean {
    const extension = path.extname(file);
    if (!this.fileExtensions.length) return true;

    // Check if the file extension is in the list of valid extensions, both match .ts or ts
    return this.fileExtensions.some(
      (ext) => ext === extension || ext === extension.slice(1),
    );
  }

  // Check if a file is a test file
  private isTestFile(file: string): boolean {
    const extension = path.extname(file);
    const testFileType = userOptions.options.testFileType;
    return file.endsWith(`.${testFileType}${extension}`);
  }

  // Get all file paths that are not test files
  public getFileResults(
    readFileType = userOptions.readFileType,
  ): IReadFileResult[] {
    if (!this.readTypeMap[readFileType])
      throw new Error('Invalid test file read type');

    const readSpinner = ora({
      text: 'ðŸš€ [huskygpt] Reading files...',
    }).start();

    try {
      const fileResults = this.readTypeMap[readFileType]().filter(
        ({ filePath: path }) =>
          path && this.hasValidExtension(path) && !this.isTestFile(path),
      );

      if (userOptions.options.debug) {
        console.log(
          '[huskygpt] read files ===>',
          fileResults.map((r) => r.filePath),
        );
      }

      fileResults.length > 0
        ? readSpinner.succeed('ðŸŒŸðŸŒŸ [huskygpt] read files successfully! ðŸŒŸðŸŒŸ')
        : readSpinner.warn('ðŸ¤”ðŸ¤” [huskygpt] read no files! ðŸ¤”ðŸ¤”');
      return fileResults;
    } catch (error) {
      readSpinner.fail(`[huskygpt] read files failed: ${error}`);
      throw error;
    }
  }
}

export default ReadFiles;
