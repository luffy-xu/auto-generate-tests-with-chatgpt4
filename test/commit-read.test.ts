import { CommitRead } from "../src/reader/commit-read";

describe('commit-read', () => {
  it('should read from git', () => {
    const commitRead = new CommitRead();
    const ret = commitRead.getFileList(__dirname + '/mock/test.diff');
    expect(ret).toMatchSnapshot();
  });
});