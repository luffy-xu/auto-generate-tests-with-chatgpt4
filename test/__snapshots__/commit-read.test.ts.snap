// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit-read should read from git 1`] = `
[
  {
    "add": 11,
    "del": 2,
    "diff": "const packageJson = JSON.parse(
   fs.readFileSync(path.join(dirname, '../package.json'), 'utf8'),
 );
 const program = new Command();
+const runTypes = ['test', 'review', 'commit'];
 
 program
   .version(packageJson.version, '-v, --version', 'output the current version')
   .description('Generate unit tests or review your code by chatgpt 4')
   .argument('<runType>', 'run type: test or review')
   .option('-k, --api-key <key>', 'Set the OpenAI API key')
   .option(
     '-t, --openai-session-token <token>',
     "OpenAI session token, 2 step to get token, If you don't set this, will use OPENAI_API_KEY, will cause fee by api key",
   )
   .option(
     '-pu, --openai-proxy-url <url>',
     'Proxy URL to use for OpenAI API requests',
   )
   .option('-m, --model <model>', 'OpenAI model to use')
   .option('-p, --prompt <prompt>', 'OpenAI prompt to use')
   .option('-mt, --max-tokens <tokens>', 'OpenAI max tokens to use')
   .option(
     '-e, --file-extensions <extensions>',
     'File extensions to read, example: .ts,.tsx',
   )
   .option(
     '-r, --read-type <type>',
     'Read files from directory or git stage, example: dir or git',
   )
   .option(
     '-s, --read-git-status <name>',
     'Read files from git stage by status default: A,R,M',
   )
   .option(
     '-d, --read-dir-name <name>',
     'Root name of the directory to read files from, example: src',
   )
   .option(
     '-f, --test-file-type <type>',
     'Generate test file type, example: test or spec',
   )
   .option(
     '-x, --test-file-extension <extension>',
     'Generate test file name extension, example: .ts or .js',
   )
   .option(
     '-n, --test-file-dir-name <name>',
     'Generate test file directory name, example: __tests__',
   )
   .option(
     '-w, --review-report-webhook <url>',
     'Webhook URL to send review report',
   )
   .action((runType, options) => {
+    if (!runTypes.includes(runType)) {
+      // exit with error
+      console.error(
+        \`Invalid run type: \${runType}, please use one of \${runTypes.join(',')}\`
+      );
+      process.exit(1);
+    }
+
     const userOptions = {
-      huskyGPTType: runType === 'test' ? 'test' : 'review',
+      huskyGPTType: runType,
       reviewTyping: options.reviewTyping,
       ...(options.apiKey && { openAIKey: options.apiKey }),
       ...(options.model && { openAIModel: options.model }),
       ...(options.prompt && { openAIPrompt: options.prompt }),
       ...(options.maxTokens && { openAIMaxTokens: Number(options.maxTokens) }),
       ...(options.fileExtensions && {
         readFileExtensions: options.fileExtensions,
       }),
       ...(options.readType && { readType: options.readType }),
       ...(options.readGitStatus && { readGitStatus: options.readGitStatus }),
       ...(options.readDirName && { readFilesRootName: options.readDirName }),
       ...(options.testFileType && { testFileType: options.testFileType }),
       ...(options.testFileExtension && {
         testFileNameExtension: options.testFileExtension,
       }),
       ...(options.testFileDirName && {
         testFileDirName: options.testFileDirName,
       }),
       ...(options.reviewReportWebhook && {
         reviewReportWebhook: options.reviewReportWebhook,
       }),
       ...(options.openAISessionToken && {
         openAISessionToken: options.openAISessionToken,
       }),
       ...(options.openAIProxyUrl && {
         openAIProxyUrl: options.openAIProxyUrl,
       }),
     };
 
     main(userOptions);
   });",
    "filename": "/bin/cli.js",
    "mode": "modify",
    "source": "/bin/cli.js",
  },
  {
    "add": 49,
    "del": 1,
    "diff": "+You are an expert programmer summarizing a git diff.
+Reminders about the git diff format:
+For every file, there are a few metadata lines, like (for example):
+\`\`\`
+diff --git a/lib/index.js b/lib/index.js
+index aadf691..bfef603 100644
+--- a/lib/index.js
++++ b/lib/index.js
+\`\`\`
+This means that \`lib/index.js\` was modified in this commit. Note that this is only an example.
+Then there is a specifier of the lines that were modified.
+A line starting with \`+\` means it was added.
+A line that starting with \`-\` means that line was deleted.
+A line that starts with neither \`+\` nor \`-\` is code given for context and better understanding.
+It is not part of the diff.
+After the git diff of the first file, there will be an empty line, and then the git diff of the next file.
+
+Do not include the file name as another part of the comment.
+Do not use the characters \`[\` or \`]\` in the summary.
+Write every summary comment in a new line.
+Comments should be in a bullet point list, each line starting with a \`-\`.
+The summary should not include comments copied from the code.
+The output should be easily readable. When in doubt, write fewer comments and not more. Do not output comments that
+simply repeat the contents of the file.
+Readability is top priority. Write only the most important comments about the diff.
+
+EXAMPLE SUMMARY COMMENTS:
+\`\`\`
+- Raise the amount of returned recordings from \`10\` to \`100\`
+- Fix a typo in the github action name
+- Move the \`octokit\` initialization to a separate file
+- Add an OpenAI API for completions
+- Lower numeric tolerance for test files
+- Add 2 tests for the inclusive string split function
+\`\`\`
+Most commits will have less comments than this examples list.
+The last comment does not include the file names,
+because there were more than two relevant files in the hypothetical commit.
+Do not include parts of the example in your summary.
+It is given only as an example of appropriate comments.
+
+
+THE GIT DIFF TO BE SUMMARIZED:
+\`\`\`
+{{ file_diff }}
+\`\`\`
+
+THE SUMMARY:",
    "filename": "/prompt/summarize_file_diff.txt",
    "mode": "add",
    "source": "/prompt/summarize_file_diff.txt",
  },
  {
    "add": 0,
    "del": 0,
    "diff": "",
    "filename": "/src/huskygpt/commit.ts",
    "mode": "add",
    "source": "/src/huskygpt/commit.ts",
  },
  {
    "add": 10,
    "del": 1,
    "diff": "import { HuskyGPTTypeEnum, IUserOptions } from './types';
 const runMap: Record<HuskyGPTTypeEnum, () => void> = {
   [HuskyGPTTypeEnum.Test]: async () => {
     const testFilePaths = new ReadFiles();
     const huskygpt = new HuskyGPTTest();
 
     // Generate a test case for each file path
     for (const fileResult of testFilePaths.getFileResults()) {
       await huskygpt.run(fileResult);
     }
   },
   [HuskyGPTTypeEnum.Review]: async () => {
     const reviewFiles = new ReadFiles();
     const huskygpt = new HuskyGPTReview();
 
     // Review code for each file path
     for (const fileResult of reviewFiles.getFileResults()) {
       await huskygpt.run(fileResult);
     }
 
     // Publish the notices to the webhook channel
     huskygpt.publishNotice();
   },
+  [HuskyGPTTypeEnum.Commit]: async () => {
+    const reviewFiles = new ReadFiles();
+    const huskygpt = new HuskyGPTReview();
+
+    // Review code for each file path
+    for (const fileResult of reviewFiles.getFileResults()) {
+      console.log(fileResult);
+    }
+  },
 };
 
 /**
  * Main function for huskygpt
  */",
    "filename": "/src/index.ts",
    "mode": "modify",
    "source": "/src/index.ts",
  },
  {
    "add": 2,
    "del": 1,
    "diff": "/**
  * The run mode of the huskygpt
  * test - Run the test file generation
  * review - Run the review file generation
  */
 export enum HuskyGPTTypeEnum {
   Test = 'test',
   Review = 'review',
+  Commit = 'commit',
 }
 
 /**
  * The file extensions to search for
  * @enum {string}
  * @property {string} Directory - Read test files from directory
  * @property {string} GitStage - Read test files from git stage
  * @readonly
  * @type {ReadTypeEnum}
  * @default 'dir'
  */",
    "filename": "/src/types.ts",
    "mode": "modify",
    "source": "/src/types.ts",
  },
  {
    "add": 2,
    "del": 2,
    "diff": "+++ b/b/b.js
-consolelog(20)
+consolelog(21)
 console.log('xx')
",
    "filename": "/b/b.js",
    "mode": "rename",
    "source": "/b/c.js",
  },
  {
    "add": 1,
    "del": 3,
    "diff": "-consolelog(20)
-console.log('haha')",
    "filename": "/b/i.js",
    "mode": "delete",
    "source": "/b/i.js",
  },
]
`;
